lastgersync/evmdownloader_fep.go

```diff
@@ -0,0 +1,210 @@
+package lastgersync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/0xPolygon/cdk-contracts-tooling/contracts/pp/l2-sovereign-chain/polygonzkevmglobalexitrootv2"
+	aggkitcommon "github.com/agglayer/aggkit/common"
+	"github.com/agglayer/aggkit/db"
+	"github.com/agglayer/aggkit/log"
+	"github.com/agglayer/aggkit/sync"
+	aggkittypes "github.com/agglayer/aggkit/types"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Event is the combination of the events that are emitted by the L2 GER manager
+type Event struct {
+	GERInfo  *GlobalExitRootInfo
+	GEREvent *GEREvent
+}
+
+type downloaderFEP struct {
+	*sync.EVMDownloaderImplementation
+	l2GERManager   *polygonzkevmglobalexitrootv2.Polygonzkevmglobalexitrootv2
+	l2GERAddr      common.Address
+	l1InfoTreeSync L1InfoTreeQuerier
+	processor      *processor
+	rh             *sync.RetryHandler
+}
+
+func newDownloaderFEP(
+	l2Client aggkittypes.BaseEthereumClienter,
+	l2GERAddr common.Address,
+	l1InfoTreeSync L1InfoTreeQuerier,
+	processor *processor,
+	rh *sync.RetryHandler,
+	blockFinality *big.Int,
+	waitForNewBlocksPeriod time.Duration,
+) (*downloaderFEP, error) {
+	l2GERManager, err := polygonzkevmglobalexitrootv2.NewPolygonzkevmglobalexitrootv2(
+		l2GERAddr, l2Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize L2 GER manager contract: %w", err)
+	}
+
+	evmDownloader := sync.NewEVMDownloaderImplementation(
+		"lastgersync", l2Client, blockFinality,
+		waitForNewBlocksPeriod, nil, nil,
+		rh, nil)
+
+	return &downloaderFEP{
+		EVMDownloaderImplementation: evmDownloader,
+		l2GERManager:                l2GERManager,
+		l2GERAddr:                   l2GERAddr,
+		l1InfoTreeSync:              l1InfoTreeSync,
+		processor:                   processor,
+		rh:                          rh,
+	}, nil
+}
+
+// RuntimeData returns the runtime data: chainID + addresses to query
+func (d *downloaderFEP) RuntimeData(ctx context.Context) (sync.RuntimeData, error) {
+	chainID, err := d.ChainID(ctx)
+	if err != nil {
+		return sync.RuntimeData{}, err
+	}
+	return sync.RuntimeData{
+		ChainID:   chainID,
+		Addresses: []common.Address{d.l2GERAddr},
+	}, nil
+}
+
+func (d *downloaderFEP) Download(ctx context.Context, fromBlock uint64, downloadedCh chan sync.EVMBlock) {
+	var (
+		attempts            int
+		nextL1InfoTreeIndex uint32
+		err                 error
+	)
+
+	// Determine the next index to start fetching GERs
+	for {
+		latestL1InfoTreeIndex, err := d.processor.getLatestL1InfoTreeIndex()
+		if errors.Is(err, db.ErrNotFound) {
+			nextL1InfoTreeIndex = 0
+		} else if err != nil {
+			log.Errorf("error getting latest l1 info tree index: %v", err)
+			attempts++
+			d.rh.Handle("getLatestL1InfoTreeIndex", attempts)
+
+			continue
+		}
+		if latestL1InfoTreeIndex > 0 {
```

Comment:
Is 0 a valid index? I am considering the case where the `getLatestL1InfoTreeIndex` returns 0 and nil as error. What happens then?

Reply:
It seems that the 0 is valid [index](https://github.com/agglayer/aggkit/blob/aa160bd90b79c2074252212341cfe547d2598eb3/l1infotreesync/processor.go#L338) for `L1InfoTreeLeaf`.

I'm trying to do a mental debugging here, so please correct me if I'm wrong...
In case `getLatestL1InfoTreeIndex` returns a 0:
- `latestL1InfoTreeIndex` is set to 0 and `nextL1InfoTreeIndex` would not be incremented by one
- therefore we would start querying GERs from 0 index, which is correct I suppose?

Did you have something else on mind? It seems, however, that we don't need both the `nextL1InfoTreeIndex` and `latestL1InfoTreeIndex`, but just one of them is enough. WDYT?

Reply:
Which is fine I guess, the l1 info tree index should be 0 based AFAIK.